### PR TITLE
Abort trends requests & fix few bugs

### LINKED ISSUE #5134: E2E query cancellation
## Is your feature request related to a problem?

I can bring a lot of misery on our servers:

It shouldn't be this simple to see a cloudflare logo when reloading PostHog. 

## Describe the solution you'd like

Proper E2E insights query cancellation. From the moment a new request is made on the frontend, the old request is disconnected and releases its load on Clickhouse. 

- We can easily abort `fetch` requests on the frontend for trends. While my dev django server says it cancels the request, will long running queries in clickhouse actually be cancelled?
- Funnel results are calculated async with Celery. How could we cancel those?

## Describe alternatives you've considered

Throttling requests on the frontend. Instead of a traditional breakpoint debounce, where you wait a certain amount of ms before performing an action, here to give the impression of speed I would always query immediately, but afterwards debounce for 0.5-1 seconds before making the next query, if any. 

## Additional context

Also, this bug is still out there: https://github.com/PostHog/posthog/issues/4858

### PR DESCRIPTION
**Pull request by mariusandra**
## Changes

Partially resolves https://github.com/PostHog/posthog/issues/5134

Partially because this is just for trends, and does not resolve funnels, paths, etc.

Before:

After:

Also fixes a bunch of random bugs:
- There was a bit of a mess with breakpoints in a few loaders. The `endQuery` function should be really called in the end, not when it fails and a new request comes in.
- There was a broken `actions.setLocalFilters({ filters: values.localFilters.filter((_, i) => i !== index) } as FilterType)` where the manual `as FilterType` was an invalid override. Let's try to avoid such things :).

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Review by liyiy — APPROVED** [review]

File: frontend/src/scenes/trends/trendsLogic.ts [hunk 1]
```diff
@@ -145,37 +145,49 @@ export const trendsLogic = kea<trendsLogicType<IndexedTrendResult, TrendResponse
...
-    loaders: ({ values, props }) => ({
+    loaders: ({ cache, values, props }) => ({
         _results: {
             __default: {} as TrendResponse,
             loadResults: async (refresh = false, breakpoint) => {
                 if (props.cachedResults && !refresh && values.filters === props.filters) {
                     return { result: props.cachedResults } as TrendResponse
                 }
+
+                // If a query is in progress, debounce for 500ms before making the second query
```
Comment (macobo) [inline frontend/src/scenes/trends/trendsLogic.ts:156]: The comment and the breakpoint don't match up.

**Comment by macobo** [comment]
> The endQuery function should be really called in the end, not when it fails and a new request comes in.

Given that `endQuery` fires some events we use to track performance can you go into a bit more depth here - what changed and how does it affect metrics?

**Comment by mariusandra** [comment]
> Given that `endQuery` fires some events we use to track performance can you go into a bit more depth here - what changed and how does it affect metrics?

Hmm. I wasn't there when this was implemented and can appreciate all the extra context.

Considering this order of events that is likely to happen now:

```
- startQuery Q1
- startQuery Q2
- endQuery Q1 (errored or aborted or breakpoint)
- endQuery Q2
```

The issue is that `endQuery` sets the reducers `maybeShowTimeoutMessage` and `maybeShowErrorMessage`. These can affect how the page behaves, potentially showing users incorrect information.

Basically the question comes down to: do we care about reporting failed queries when the user has already started a new query and will never see the failed results? If not, the PR should be good as is. If yes, should I also report aborted queries separately somehow?

**Comment by macobo** [comment]
> do we care about reporting failed queries when the user has already started a new query and will never see the failed results?

I think this is good as-is though capturing aborted queries might be valuable on it's own right as a separate thing.

**Comment by mariusandra** [comment]
I added reporting for aborts: https://github.com/PostHog/posthog/pull/5136/files#diff-db6bd5dcc4a5a5da668472f1e183b343018401fcfa5390c57b06487820de900fR170-R181

Do all the keys, etc look good?

Apologies for the noise in the PR. I had to improve some types, and in the end chose the least horrible refactor that just reshuffled a few imports...

**Comment by macobo** [comment]
LGTM, solid stuff. Probably need to adjust some things long-term but we should ship this. :)

**Comment by mariusandra** [comment]
Ah, no, a bug. One ".js" file is using one TypeScript enum and getting "undefined". I'll fix that and then merge when tests pass.

**Comment by mariusandra** [comment]
Well, almost. Using TypeScript enums in JS apparently works! It just wasn't smart enough to know that the import path had changed.

**Pull request merged**